echopype/convert/parse_base.py

```diff
@@ -69,54 +76,274 @@ def _print_status(self):
             f"parsing file {os.path.basename(self.source_file)}, " f"time of first ping: {time}"
         )
 
-    def rectangularize_data(self):
+    @property
+    def num_transducer_sectors(self) -> Dict[Any, int]:
+        """Get the number of transducer sectors for each channel.
+        This is for receive raw type only."""
+        num_sectors = {}
+        if self.ping_data_dict["n_complex"]:
+            n_complex = self.ping_data_dict["n_complex"]
+            for ch_id, n_complex_list in n_complex.items():
+                num_transducer_sectors = np.unique(np.array(n_complex_list))
+                if num_transducer_sectors.size > 1:  # this is not supposed to happen
+                    raise ValueError("Transducer sector number changes in the middle of the file!")
+                else:
+                    num_transducer_sectors = num_transducer_sectors[0]
+
+                num_sectors[ch_id] = int(num_transducer_sectors)
+        return num_sectors
+
+    def _get_data_shapes(self) -> dict:
+        """Get all the expanded data shapes"""
+        all_data_shapes = {}
+        # Get all data type shapes
+        for raw_type in self.raw_types:
+            ping_data_dict = (
+                self.ping_data_dict_tx if raw_type == "transmit" else self.ping_data_dict
+            )
+            data_type_shapes = calc_final_shapes(self.data_types, ping_data_dict)
```

Comment:
Just adding some comments for readability as I go through the code
```suggestion
            # data_types: ["power", "angle", "complex"]
            data_type_shapes = calc_final_shapes(self.data_types, ping_data_dict)
```